### PR TITLE
[inductor] online softmax

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127011

Softmax need do some preparation work that access the input tensor in two passes
- compute amax of each row
- compute (x - amax).exp.sum for each row

When the row size is large, cache can not hold all the active data and accessing the input multiple passes increases execution time since the kernel is membw bounded.

Online softmax uses a customized reduction to compute max and sum at the same time by accessing the data in one pass. Check this paper for more details ( https://arxiv.org/abs/1805.02867 ). 

Also here is an online softmax kernel generated by inductor as a reference: https://gist.github.com/shunting314/67ae4fffd45d4f2753c781780332fa54

## Microbenchmark

- `TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_ONLINE_SOFTMAX=0 DO_PERF_TEST=1 python test/inductor/test_online_softmax.py -k test_softmax` : without online softmax
  - eager_ms=6.671296119689941
  - opt_ms=8.06931209564209
- `TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_ONLINE_SOFTMAX=1 DO_PERF_TEST=1 python test/inductor/test_online_softmax.py -k test_softmax`: with online softmax
  - eager_ms=6.634047985076904
  - opt_ms=6.230591773986816

Ideally, online softmax should save about 2ms here. We saves about 1.84ms in practice.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @peterbell10 @ngimel

### CODE REVIEW

File: torch/_inductor/inductor_prims.py [hunk 1]
```diff
@@ -100,6 +105,12 @@ def eager_force_stride(input_tensor: Tensor, stride) -> Tensor:
     lambda a, b, c: (a * b) + c,
     doc="Fused multiply add: fma(a, b, c) -> (a * b) + c without rounding after the multiplication",
 )
+online_softmax = make_prim(
```
Comment (Chillee) [inline torch/_inductor/inductor_prims.py:108]: I wouldn't call this `online_softmax` - I'd call it like `compute_softmax_statistics` or something. it's not actually computing softmax.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1326,8 +1327,79 @@ def create_multilayer_existing_ranges(
...
+            inner_fn,
+            ranges,
+            reduction_ranges,
+            reduction_type,
+            src_dtype,
+            reduction_hint,
+        )
+        self.output_index = output_index
+
+    def store_reduction(self, output_name, indexer, vars, reduction_vars):
```
Comment (Chillee) [inline torch/_inductor/ir.py:1357]: This is just creating multiple identical reductions and indexing the output? Can we make welfordreduction also use this?

**Review by Chillee — COMMENTED** [review]
Please check the numerical accuracy of online softmax vs. regular softmax. Like, compute the difference vs. a fp64 reference.

**Review by jansel — APPROVED** [review]

**Review by Chillee — CHANGES_REQUESTED** [review]
I'm definitely worried about numerical accuracy here.

**Comment by shunting314** [comment]
> I'm definitely worried about numerical accuracy here.

I'll definitely do the test you mentioned before merge :)

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -741,3 +741,76 @@ def max_pool2d_with_indices(
...
+
+    # Don't do online softmax for scalar or 1d tensor
+    if x.dim() < 2:
+        return False
+
+    # Only do online softmax for GPU for now
+    if x.device.type != "cuda":
+        return False
+
+    return x.size(dim) > 2**14
```
Comment (eellison) [inline torch/_inductor/decomposition.py:758]: This is going to add a guard.. is that we want ?

It's also going to cause all other pattern matching that involves softmax to fail after x.size(dim) > 2 ** 14 to fail, such as SDPA.

Can we instead do this as a pattern match that runs after other passes ?

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -741,3 +741,76 @@ def max_pool2d_with_indices(
...
+        return False
+
+    # Don't do online softmax for scalar or 1d tensor
+    if x.dim() < 2:
+        return False
+
+    # Only do online softmax for GPU for now
+    if x.device.type != "cuda":
+        return False
+
```
Comment (eellison) [inline torch/_inductor/decomposition.py:757]: add comment why this particular number?  could we also make this a config option so that we can test it more easily on lower sizes ?

**Review by eellison — COMMENTED** [review]
I think it would be better to do this as a late pattern matching than in decomposition. Would also be great to have more accuracy tests that move the config down to a lower threshold.

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (shunting314) [reply torch/_inductor/decomposition.py:758]: > This is going to add a guard.. is that we want ?

Good point. Any suggestion to improve it?

Comment (shunting314) [reply torch/_inductor/decomposition.py:758]: > It's also going to cause all other pattern matching that involves softmax to fail after x.size(dim) > 2 ** 14 to fail, such as SDPA.

I give this another thought. But even if the online softmax decomposition rule kicks in for SDPA, this is still fine? Because pattern should also generated Fx graphs with online softmax. This does not affect if the pattern can get a match or not. Wdyt?

I can also add some tests for this.

Comment (shunting314) [reply torch/_inductor/decomposition.py:758]: hmm, the example inputs to trace the pattern is quite small (not the same as the inputs the the softmax in the model will get). That may cause some diecrepancy.

Comment (Chillee) [reply torch/_inductor/decomposition.py:758]: How does the performance compare on smaller shapes?

Comment (shunting314) [reply torch/_inductor/decomposition.py:758]: > How does the performance compare on smaller shapes? 

May I know why we care about perf on smaller shapes?

The 2 major usages of softmax I care most are:
1. those used in the attention layer, which we should leverage fused attention like flash attention
2. the softmax inside CrossEntryLoss. In which case, the reduction dimension is the vocab size which should most likely be larger than the 16K threshold here.

Do I miss some other important cases?

Comment (Chillee) [reply torch/_inductor/decomposition.py:758]: If it works well on smaller cases then we can just always turn it on.

Comment (eellison) [reply torch/_inductor/decomposition.py:758]: > I give this another thought. But even if the online softmax decomposition rule kicks in for SDPA, this is still fine? Because the pattern should also generated Fx graphs with online softmax. This does not affect if the pattern can get a match or not. Wdyt?

The difficulty is that it only sometimes fires.

Comment (eellison) [reply torch/_inductor/decomposition.py:758]: Do we need to enable persistent kernel/split reduction to always turn it on ? 

Also, doing a late pattern is not too difficult. it might be an easier first step to landing

Comment (eellison) [reply torch/_inductor/decomposition.py:758]: Re: size. you could just disable if there are symints. or there is a suppress_guards api

File: test/inductor/test_online_softmax.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        res_error = rmse(ref_fp64[1], res[1]).item()
+
+        # My local tests even shows a smaller res_error:
+        #   ref_error=2.1065, res_error=2.1028
+        # for bf16
+        #   ref_error=0.2611, res_error=0.2609
+        # for fp16
+        #   ref_error=0.0001, res_error=0.0001
+        # for fp32
+        print(f"{ref_error=:.4f}, {res_error=:.4f}")
```
Comment (shunting314) [inline test/inductor/test_online_softmax.py:184]: @Chillee the numerics of online softmax is even slightly better in my tests compared to a fp64 baseline

**Review by jansel — CHANGES_REQUESTED** [review]
failing tests?

**Comment by shunting314** [comment]
The accuracy failure for speech_transformer is real and it's due to enabling online softmax. I'll need debug more.

**Comment by shunting314** [comment]
@jansel @eellison this is ready for review again.

Benchmarked for llm.c 
- without online softmax or fast math: 175.163ms , 187.1K tok/s 
- with online softmax and fast math: 173.184ms, 189.2K tok/s

A100 dashboard is [neutral](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Tue%2C%2025%20Feb%202025%2018%3A30%3A42%20GMT&stopTime=Tue%2C%2004%20Mar%202025%2018%3A30%3A42%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/shunting314/151/head&lCommit=4f6748484afaf9cc736f560564f73a85525be2c6&rBranch=main&rCommit=dae3fbfe9720e83e7e81d41430fb5067221bbed7). The tensors are much smaller so gain is also smaller

H100 dashboard shows 6% speedup for TB & 3% speedup for HF [link](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Tue%2C%2025%20Feb%202025%2018%3A30%3A42%20GMT&stopTime=Tue%2C%2004%20Mar%202025%2018%3A30%3A42%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/shunting314/151/head&lCommit=4f6748484afaf9cc736f560564f73a85525be2c6&rBranch=main&rCommit=dae3fbfe9720e83e7e81d41430fb5067221bbed7). Although I think these are just noise.

**Review by jansel — APPROVED** [review]
